src/Java.Interop.Tools.JavaTypeSystem/Adapters/ManagedApiImporter.cs

```diff
@@ -0,0 +1,379 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Java.Interop.Tools.Cecil;
+using Java.Interop.Tools.JavaTypeSystem.Models;
+using Mono.Cecil;
+using Mono.Collections.Generic;
+
+namespace Java.Interop.Tools.JavaTypeSystem
+{
+	public static class ManagedApiImporter
+	{
+		public static JavaTypeCollection Parse (AssemblyDefinition assembly, JavaTypeCollection collection)
+		{
+			var types_to_add = new List<JavaTypeModel> ();
+
+			foreach (var md in assembly.Modules)
+				foreach (var td in md.Types) {
+					if (!ShouldSkipType (td) && ParseType (td, collection) is JavaTypeModel type)
+						types_to_add.Add (type);
+				}
+
+			// This needs to be done ordered from least nested to most nested, in order for nesting to work.
+			// That is, 'android.foo.Blah' needs to be added before 'android.foo.Blah.Bar'.
+			// Plus, we may have unnested managed types that are actually nested in Java-land:
+			// ex: IContextMenu and IContextMenuContextMenuItem
+			foreach (var type in types_to_add.OrderBy (t => t.FullName.Count (c => c == '.')).ToArray ())
+				AddReferenceTypeRecursive (type, collection);
+
+			return collection;
+		}
+
+		public static JavaTypeModel? ParseType (TypeDefinition type, JavaTypeCollection collection)
+		{
+			if (!type.IsPublic && !type.IsNested)
+				return null;
+
+			if (!ShouldImport (type))
+				return null;
+
+			var model = type.IsInterface ? (JavaTypeModel?) ParseInterface (type, collection) : ParseClass (type, collection);
+
+			if (model is null)
+				return null;
+
+			foreach (var nested in type.NestedTypes)
+				if (ParseType (nested, collection) is JavaTypeModel nested_model)
+					model.NestedTypes.Add (nested_model);
+
+			return model;
+		}
+
+		static bool ShouldImport (TypeDefinition td)
+		{
+			// We want to exclude "IBlahInvoker" and "IBlahImplementor" and "BlahConsts" types
+			if (td.Name.EndsWith ("Invoker")) {
+				var n = td.FullName;
+				n = n.Substring (0, n.Length - 7);
+
+				var types = td.DeclaringType != null ? td.DeclaringType.Resolve ().NestedTypes : td.Module.Types;
+
+				if (types.Any (t => t.FullName == n))
+					return false;
+			}
+
+			if (td.Name.EndsWith ("Implementor")) {
+				var n = td.FullName;
+				n = n.Substring (0, n.Length - 11);
+
+				var types = td.DeclaringType != null ? td.DeclaringType.Resolve ().NestedTypes : td.Module.Types;
+
+				if (types.Any (t => t.FullName == n))
+					return false;
+			}
+
+			if (td.Name.EndsWith ("Consts")) {
+				var n = td.FullName;
+				n = n.Substring (0, n.Length - 6);
+
+				var types = td.DeclaringType != null ? td.DeclaringType.Resolve ().NestedTypes : td.Module.Types;
+
+				if (types.Any (t => t.FullName == n))
+					return false;
+			}
+
+			return true;
+		}
+
+		public static JavaClassModel? ParseClass (TypeDefinition type, JavaTypeCollection collection)
+		{
+			// TODO: type parameters?
```

Comment:
We could `Debug.Assert()` that `type` has no generic type parameters for now.

Reply:
…plus if/when we ever support [`[JavaTypeArguments]`](https://github.com/xamarin/java.interop/issues/590#issuecomment-648364368), that could be "forwarded" here as well.

Reply:
I **think** type parameter data already exists in the `api.xml` emitted by `class-parse`, we are just not handling it in `generator`.  Thus it would assert.